h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,185 @@
+package ai.h2o.automl.targetencoding;
+
+import water.MRTask;
+import water.fvec.CategoricalWrappedVec;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.util.DistributedException;
+import water.util.Log;
+
+class BroadcastJoinForTargetEncoder {
+
+  private static String FOLD_VALUE_IS_OUT_OF_RANGE_MSG = "Fold value should be a non-negative integer (i.e. should belong to [0, Integer.MAX_VALUE] range)";
+
+
+  static class FrameWithEncodingDataToArray extends MRTask<FrameWithEncodingDataToArray> {
+    
+    /**
+     * As it is less likely that we have more folds than categorical levels It is more preferable to keep `num` and `den` in a separate (but adjacent) rows of the 2D array and not in the same row with shifting.
+     *     For _foldColumnIdx == -1
+     *        _encodingDataPerNode[0] will be storing numerators
+     *        _encodingDataPerNode[1] will be storing denominators
+     *     For _foldColumnIdx != -1
+     *        _encodingDataPerNode[2k-2] will be storing numerators, 
+     *        _encodingDataPerNode[2k-1] will be storing denominators, where k is a current fold ; folds = {1,2,3...k}
+     */
+    int[][] _encodingDataPerNode = null;
+    
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx;
+
+    FrameWithEncodingDataToArray(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx, int cardinalityOfCatCol, int numberOfFolds) {
+      _categoricalColumnIdx = categoricalColumnIdx;
+      _foldColumnIdx = foldColumnId;
+      _numeratorIdx = numeratorIdx;
+      _denominatorIdx = denominatorIdx;
+      if(foldColumnId == -1) {
+        _encodingDataPerNode = new int[2][cardinalityOfCatCol]; 
+      } else {
+        assert numberOfFolds > 0 : "Number of folds should be greater than zero";
+        assert cardinalityOfCatCol > 0 : "Cardinality of categ. column should be greater than zero";
+        _encodingDataPerNode = new int[numberOfFolds * 2][cardinalityOfCatCol];
+      }
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        int levelValue = (int) categoricalChunk.at8(i);
+        int[] arrForNumerators = null;
+        int[] arrForDenominators = null;
+
+        if (_foldColumnIdx != -1) {
+          long foldValueFromVec = cs[_foldColumnIdx].at8(i);
+          if (!(foldValueFromVec <= Integer.MAX_VALUE && foldValueFromVec > 0)) {
+            throw new DistributedException(new AssertionError(FOLD_VALUE_IS_OUT_OF_RANGE_MSG + " but was " + foldValueFromVec));
```

Comment:
use `assert` instead

also: this doesn't have to be checked for every single row - just check the min-max before you actually submit the task